### PR TITLE
feat: Implement CoreCrypto file locks - FS-1557 

### PR DESCRIPTION
**Pull request by netbe**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

* Create a SafeCoreCrypto class that will handle the lock on the database for CoreCrypto
* Add TODO to restoreFromDisk directly after acquiring lock.

#### Test Coverage (Optional)

Could not create a test to test flock from multiple processes.

#### How to Test

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-data-model/Source/MLS/SafeCoreCrypto.swift [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    func perform<T>(_ block: (CoreCryptoProtocol) throws -> T) rethrows -> T
+    func unsafePerform<T>(_ block: (CoreCryptoProtocol) throws -> T) rethrows -> T
+}
+
+public struct SafeCoreCrypto: SafeCoreCryptoProtocol {
+
+    private let coreCrypto: CoreCryptoProtocol
+    private let safeContext: SafeFileContext
+
+    init(coreCrypto: CoreCryptoProtocol, coreCryptoConfiguration: CoreCryptoConfiguration) {
```
Comment (johnxnguyen) [inline wire-ios-data-model/Source/MLS/SafeCoreCrypto.swift:51]: Perhaps we can have a convenience init that will create the instance of `CoreCrypto` using the configuration, what do you think?

File: wire-ios-data-model/Source/MLS/SafeCoreCrypto.swift [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        self.coreCrypto = coreCrypto
+        let directoryPathUrl = URL(fileURLWithPath: coreCryptoConfiguration.path).deletingLastPathComponent()
+        self.safeContext = SafeFileContext(fileURL: directoryPathUrl)
+    }
+
+    public func perform<T>(_ block: (CoreCryptoProtocol) throws -> T) rethrows -> T {
+        var result: T
+        safeContext.acquireDirectoryLock()
```
Comment (johnxnguyen) [inline wire-ios-data-model/Source/MLS/SafeCoreCrypto.swift:59]: Is this blocking?

Comment (netbe) [reply wire-ios-data-model/Source/MLS/SafeCoreCrypto.swift:59]: nope, so the issue was I was setting the lock on the db file and not the directory

File: wire-ios-data-model/Source/MLS/SafeCoreCrypto.swift (hunk 1 above)
Comment (netbe) [reply wire-ios-data-model/Source/MLS/SafeCoreCrypto.swift:51]: yep I can do it

Comment (netbe) [reply wire-ios-data-model/Source/MLS/SafeCoreCrypto.swift:51]: that would remove the `CoreCryptoFactory.createCoreCrypto` method

Comment (johnxnguyen) [reply wire-ios-data-model/Source/MLS/SafeCoreCrypto.swift:51]: Sure, or we can wrap the init of `SafeCoreCrypto` inside the factory.

File: wire-ios-data-model/Source/MLS/SafeCoreCrypto.swift (hunk 2 above)
Comment (johnxnguyen) [reply wire-ios-data-model/Source/MLS/SafeCoreCrypto.swift:59]: I'm a bit confused, not sure what the issue you're referring to is. I mean, does execution stop on the current thread until the lock is acquired?

Comment (netbe) [reply wire-ios-data-model/Source/MLS/SafeCoreCrypto.swift:59]: ah sorry that was another issue I had... I didn't really notice

File: wire-ios-data-model/Source/Proteus/ProteusService.swift [hunk 3]
```diff
@@ -26,14 +26,15 @@ public final class ProteusService: ProteusServiceInterface {
...
     private let logger = WireLogger.proteus
 
     // MARK: - Life cycle
 
-    public init(coreCrypto: CoreCryptoProtocol) throws {
+    public init(coreCrypto: SafeCoreCryptoProtocol) throws {
         self.coreCrypto = coreCrypto
-        try coreCrypto.proteusInit()
+        // this is fine not to lock for initialization
+        try coreCrypto.perform { try $0.proteusInit() }
```
Comment (johnxnguyen) [inline wire-ios-data-model/Source/Proteus/ProteusService.swift:36]: Should this be unsafe perform then?

**Review by johnxnguyen — APPROVED** [review]

File: wire-ios-data-model/Source/Proteus/ProteusService.swift (hunk 3 above)
Comment (netbe) [reply wire-ios-data-model/Source/Proteus/ProteusService.swift:36]: no it's fine too

File: wire-ios-data-model/Source/MLS/MLSController.swift [hunk 4]
```diff
@@ -581,10 +583,7 @@ public final class MLSController: MLSControllerProtocol {
         }
 
         do {
-            let groupIDBytes = try coreCrypto.processWelcomeMessage(
-                welcomeMessage: messageBytes,
-                customConfiguration: .init(keyRotationSpan: nil, wirePolicy: nil)
-            )
+            let groupIDBytes = try coreCrypto.perform { try $0.processWelcomeMessage(welcomeMessage: messageBytes, customConfiguration: .init(keyRotationSpan: nil, wirePolicy: nil)) }
```
Comment (David-Henner) [inline wire-ios-data-model/Source/MLS/MLSController.swift]:
```suggestion
            let groupIDBytes = try coreCrypto.perform { 
                try $0.processWelcomeMessage(
                    welcomeMessage: messageBytes, 
                    customConfiguration: .init(keyRotationSpan: nil, wirePolicy: nil)
                )
             }
```

File: wire-ios-data-model/Source/MLS/SafeCoreCrypto.swift [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        // TODO: call `restoreFromDisk`
```
Comment (David-Henner) [inline wire-ios-data-model/Source/MLS/SafeCoreCrypto.swift]: what's preventing us from calling `restoreFromDisk` already?

File: wire-ios-data-model/Tests/MLS/SafeCoreCryptoTests.swift [hunk 6]
```diff
@@ -0,0 +1,41 @@
+//
+//  SafeCoreCryptoTests.swift
+//  WireDataModelTests
+//
+//  Created by F on 22/02/2023.
+//  Copyright © 2023 Wire Swiss GmbH. All rights reserved.
+//
```
Comment (David-Henner) [inline wire-ios-data-model/Tests/MLS/SafeCoreCryptoTests.swift]: let's add the Wire header here

File: wire-ios-data-model/Tests/MLS/SafeCoreCryptoTests.swift [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+import Foundation
+@testable import WireDataModel
+
+func createTempFolder() -> URL {
+    let url = URL(fileURLWithPath: [NSTemporaryDirectory(), UUID().uuidString].joined(separator: "/"))
+    try! FileManager.default.createDirectory(at: url, withIntermediateDirectories: true, attributes: [:])
+    return url
+}
```
Comment (David-Henner) [inline wire-ios-data-model/Tests/MLS/SafeCoreCryptoTests.swift]: why is this outside of the test class?

File: wire-ios-data-model/Tests/MLS/SafeCoreCryptoTests.swift [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    func test_safeFileContext() throws {
+        let tempURL = createTempFolder()
+        let sut = SafeFileContext(fileURL: tempURL)
+
+        sut.acquireDirectoryLock()
+        print("locked")
+        sut.releaseDirectoryLock()
+        print("unlocked")
+
+    }
```
Comment (David-Henner) [inline wire-ios-data-model/Tests/MLS/SafeCoreCryptoTests.swift]: what makes this test successful? not timing out?

**Review by David-Henner — COMMENTED** [review]
there's an empty file at `wire-ios-share-engine/Sources/MLS/MLSEncryptionController.swift` and `wire-ios-sync-engine/Source/UserSession/ZMUserSession+MLS.swift`

File: wire-ios-data-model/Source/MLS/SafeCoreCrypto.swift (hunk 5 above)
Comment (netbe) [reply wire-ios-data-model/Source/MLS/SafeCoreCrypto.swift]: fixed

File: wire-ios-data-model/Tests/MLS/SafeCoreCryptoTests.swift (hunk 8 above)
Comment (netbe) [reply wire-ios-data-model/Tests/MLS/SafeCoreCryptoTests.swift]: removed as there's nothing to assert

**Review by David-Henner — APPROVED** [review]

**Pull request merged**